torch/csrc/jit/codegen/cuda/arith.cpp

```diff
@@ -33,33 +31,84 @@ TORCH_CUDA_API Val* newValLike(const Val* const val, DataType dtype) {
 
   TORCH_CHECK(
       false,
-      "Could not generate a new value of type ",
-      val->getValType().value(),
-      " with data type ",
-      val->getDataType().value());
+      "Was expecting a scalar type, but received ValType: ",
+      vtype,
+      " with DataType:",
+      dtype);
 }
 
-TORCH_CUDA_API Val* newValLike(const Val* const val) {
-  return newValLike(val, val->getDataType().value());
+TensorView* newOutputTV(const std::vector<Val*>& vals, DataType dtype) {
+  std::vector<TensorView*> tvs;
+  for (auto val : vals)
+    if (val->getValType() == ValType::TensorView)
+      tvs.push_back(static_cast<TensorView*>(val));
+
+  TORCH_CHECK(
+      !tvs.empty(),
+      "Tried to create new output TensorView but received empty list.");
+
+  std::vector<IterDomain*> out_domain(
+      tvs[0]->domain()->noReductions().size(), nullptr);
+
+  for (auto tv : tvs) {
+    auto dom = tv->domain()->noReductions();
+    TORCH_INTERNAL_ASSERT(
+        dom.size() == out_domain.size(),
+        "Invalid tensor view found while producing and output, it has ",
+        dom.size(),
+        " dimensions but expected ",
+        out_domain.size());
+    for (size_t i = 0; i < dom.size(); i++) {
+      if (out_domain[i] != nullptr)
+        continue;
+      if (dom[i]->isBroadcast())
+        continue;
+      out_domain[i] = new IterDomain(dom[i]->start(), dom[i]->extent());
+    }
+  }
+
+  // If any dims are missing, fill as broadcast dim.
+  for (size_t i = 0; i < out_domain.size(); i++)
+    if (out_domain[i] == nullptr)
+      out_domain[i] = new IterDomain(
+          new Int(0), new Int(1), ParallelType::Serial, false, false, true);
+
+  return new TensorView(new TensorDomain(out_domain), dtype);
+}
+
+Val* newOutputVal(const std::vector<Val*>& vals) {
+  TORCH_INTERNAL_ASSERT(
+      !vals.empty(), "Cannot promote values if there aren't any.");
+
+  ValType out_vtype = vals[0]->getValType().value();
+  DataType out_dtype = vals[0]->getDataType().value();
+
+  for (auto val : vals) {
+    TORCH_CHECK(val->isVal(), "Invalid statement found during promotion.");
+    TORCH_CHECK(
+        val->getDataType().value() != DataType::Null,
+        "Invalid datatype found during prmotion.");
+    out_vtype = promote_type(out_vtype, val->getValType().value());
+    out_dtype = promote_type(out_dtype, val->getDataType().value());
```

Comment:
This is the place where I was asking earlier.
out_dtype promotion should be weighted by ValType of the input.

Reply:
Please ignore the previous comment. This will require some further investigation to figure out the proper way to handle promotion. But it's not a blocking issue.